Cubical/Algebra/CONTRIBUTING.md

```diff
@@ -0,0 +1,91 @@
+  2. Properties of the carrier: this is usually `isSet`.
+  3. Operations: Non-propositional operations on the carrier, such as the multiplication and unit for a monoid.
+  4. Properties of the operations: for example, the associativity and unit laws for a monoid.
+
+  We separate (ii) and (iv) because it makes it easier to avoid redundant assumptions of `isSet` in complex
+  algebraic structures or to drop this assumption when it is not needed. As an example illustrating our naming
+  conventions, we would give the following definitions building up the universe of monoids.
+
+  ```
+  RawMonoidStr : Type ℓ → Type ℓ
+  RawMonoidStr = {! unit and multiplication operators !}
+
+  IsMonoidStr : (A : Type ℓ) → RawMonoidStr A → Type ℓ
+  IsMonoidStr = {! associativity and unit laws !}
+
+  MonoidStr : Type ℓ → Type ℓ
+  MonoidStr A = Σ[ R ∈ RawMonoidStr A ] (IsMonoidStr R)
+
+  Monoid : Type (ℓ-suc ℓ)
+  Monoid = Σ[ A ∈ Type ℓ ] (isSet A × MonoidStr A)
+  ```
+
+  _Is this definition of `Monoid` best, or should it be rearranged to `Σ[ A ∈ hSet ℓ ] MonoidStr (A .fst)`?
+  Should we use a record instead of this four-part `Σ`-type? I'd guess we want sub-groupings of these often
+  enough that a `Σ`-type is preferable._
+
+- We define more complicated instances of (i-iv) by nesting simpler instances. For example, the definition of
+  `RawRingStr` structure would be a record or `Σ`-type containing `RawMonoidStr` and `RawAbGroupStr` as
+  fields, and `IsRingStr` would likewise use `IsMonoidStr` and `IsAbGroupStr`.
+
+- We prefer records over nested `Σ`-types for efficiency and readability. For example, the definition of
+  `IsRingStr` could be a record with first `IsMonoidStr` and `IsAbGroupStr` fields and then fields for the
+  laws concerning both the monoid and group operations. Reflection
+  ([`Cubical.Reflection.RecordEquiv`](https://github.com/agda/cubical/blob/master/Cubical/Reflection/RecordEquiv.agda))
+  can be used to derive an equivalent `Σ`-type representation when this is more convenient, for example to
+  prove that some record type is propositional.
+
+- When defining an instance of an algebraic structure, use copattern matching rather than a record
+  constructor, and try to avoid relying on helper functions that take many arguments---this is for efficiency
+  reasons. Because of the heavily nested nature of algebraic structures, it can be more readable to separate
+  out components into separate definitions (local or otherwise), for example defining the group component of a
+  given ring structure separately.
+
+- It is preferred to include redundant structure and properties if there are often more direct proofs in
+  specific instances or if there is no canonical choice of minimal axioms. Helper functions can then be
+  defined to generate the full structure from a subset. For example, the definition of groups includes both a
+  left inverse and right inverse law although this is not strictly necessary. Use your judgment.
```

Comment:
What was the status of default definitions for fields?

Reply:
It looks like this is possible to some extent with reflection, I'll try and figure out how it works.

Reply:
Something that makes this hard to do cleanly with reflection is that we have multiple incompatible ways we want to infer the full structure. For example, we want to infer the left inverse law of a group if we have the right inverse, but also infer the right inverse if we have the left inverse. It seems there's no way to order the pieces defining the record so that both are possible, although we could get one of them.

We could do it by in a more complicated way: have a field `oneInverse : LeftInverse ⊎ RightInverse` and then optional fields `linv : LeftInverse` and `rinv : RightInverse` that are filled in automatically depending on what's given for `oneInverse`. But I suspect this would make other things a pain.